### PR TITLE
feat(rust): add bacon-ls and improve debugging

### PR DESCRIPTION
**Pull request by crisidev**
## Changes

**NOTE: the PR will be in draft until https://github.com/mason-org/mason-registry/pull/5774 is not merged**

Some improvements in the Rust extra

- Set `codelldb` adapter for rustacean.nvim
- Add support for [bacon-ls](https://github.com/crisidev/bacon-ls) ([blog post](https://lmno.lol/crisidev/bacon-language-server))

`bacon-ls` can be used as an alternative to `rust-analyzer` for diagnostics, improving `rust-analyzer` performances. This is configured by `vim.g.lazyvim_rust_diagnostics`, which can be set to `rust-analyzer` or `bacon-lsp`.

screenshot

### CODE REVIEW

**Comment by mirsella** [comment]
i just discovered bacon, and it looks awesome, but i don't think lazyvim should swap the default rust LSP everyone is used to, and assume is used ?
If I understand correctly, bacon runs alone without rust analyzer, or both runs at the same time ?

maybe as a extra extra ? like rust-bacon ?

**Comment by mirsella** [comment]
Thanks for the explanation.
i tried your rust.lua (even merged it to latest to test), but for me, I get no diagnostic whatsoever. bacon-ls is running alongside rust-analyzer, but nothing give any diagnostics

**Comment by dpetka2001** [comment]
I also tried your PR and don't get any diagnostics from `bacon` compared to the default LazyVim implementation. This is the configuration I used for `rust.lua` after rebasing to latest LazyVim. I don't get the error that @mirsella mentioned.

I left out the rust-analyzer stuff in `nvim-lspconfig` spec (even though this PR won't install it because of `rust_analyzer.enabled = false`), because as already discussed in #3389 and #2755 it was decided that users would be better off installing rust-analyzer via their Rust toolchain when using `rustaceanvim`.
<details>

```lua
if lazyvim_docs then
  -- LSP Server to use for Rust.
  -- Set to "bacon-ls" to use bacon-ls instead of rust-analyzer.
  -- only for diagnostics. The rest of LSP support will still be
  -- provided by rust-analyzer.
  vim.g.lazyvim_rust_diagnostics = "rust-analyzer"
end
local diagnostics = vim.g.lazyvim_rust_diagnostics or "rust-analyzer"

return {
  recommended = function()
    return LazyVim.extras.wants({
      ft = "rust",
      root = { "Cargo.toml", "rust-project.json" },
    })
  end,

  -- Extend auto completion
  {
    "hrsh7th/nvim-cmp",
    dependencies = {
      {
        "Saecki/crates.nvim",
        event = { "BufRead Cargo.toml" },
        opts = {
          completion = {
            cmp = { enabled = true },
          },
        },
      },
    },
    ---@param opts cmp.ConfigSchema
    opts = function(_, opts)
      opts.sources = opts.sources or {}
      table.insert(opts.sources, { name = "crates" })
    end,
  },

  -- Add Rust & related to treesitter
  {
    "nvim-treesitter/nvim-treesitter",
    opts = { ensure_installed = { "rust", "ron" } },
  },

  -- Ensure Rust debugger is installed
  {
    "williamboman/mason.nvim",
    optional = true,
    opts = function(_, opts)
      opts.ensure_installed = opts.ensure_installed or {}
      vim.list_extend(opts.ensure_installed, { "codelldb" })
      if diagnostics == "bacon-ls" then
        vim.list_extend(opts.ensure_installed, { "bacon", "bacon-ls" })
      end
    end,
  },

  {
    "mrcjkb/rustaceanvim",
    version = "^4", -- Recommended
    ft = { "rust" },
    opts = {
      server = {
        on_attach = function(_, bufnr)
          vim.keymap.set("n", "<leader>cR", function()
            vim.cmd.RustLsp("codeAction")
          end, { desc = "Code Action", buffer = bufnr })
          vim.keymap.set("n", "<leader>dr", function()
            vim.cmd.RustLsp("debuggables")
          end, { desc = "Rust Debuggables", buffer = bufnr })
        end,
        default_settings = {
          -- rust-analyzer language server configuration
          ["rust-analyzer"] = {
            cargo = {
              allFeatures = true,
              loadOutDirsFromCheck = true,
              buildScripts = {
                enable = true,
              },
            },
            -- Enable diagnostics if using rust-analyzer
            diagnostics = {
              enable = diagnostics == "rust-analyzer",
            },
            -- Add clippy lints for Rust.
            checkOnSave = {
              enable = diagnostics == "rust-analyzer",
              allFeatures = true,
              command = "clippy",
              extraArgs = { "--no-deps" },
            },
            procMacro = {
              enable = true,
              ignored = {
                ["async-trait"] = { "async_trait" },
                ["napi-derive"] = { "napi" },
                ["async-recursion"] = { "async_recursion" },
              },
            },
          },
        },
      },
    },
    config = function(_, opts)
      -- Configure dap with codelldb
      local package_path = require("mason-registry").get_package("codelldb"):get_install_path()
      opts.dap = {
        adapter = require("rustaceanvim.config").get_codelldb_adapter(
          package_path .. "/codelldb",
          package_path .. "/extension/lldb/lib/liblldb.dylib"
        ),
      }
      vim.g.rustaceanvim = vim.tbl_deep_extend("keep", vim.g.rustaceanvim or {}, opts or {})
      if vim.fn.executable("rust-analyzer") == 0 then
        LazyVim.error(
          "**rust-analyzer** not found in PATH, please install it.\nhttps://rust-analyzer.github.io/",
          { title = "rustaceanvim" }
        )
      end
    end,
  },

  -- Correctly setup lspconfig for Rust 🚀
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        bacon_ls = {
          enabled = diagnostics == "bacon-ls",
        },
        taplo = {
          keys = {
            {
              "K",
              function()
                if vim.fn.expand("%:t") == "Cargo.toml" and require("crates").popup_available() then
                  require("crates").show_popup()
                else
                  vim.lsp.buf.hover()
                end
              end,
              desc = "Show Crate Documentation",
            },
          },
        },
      },
    },
  },

  {
    "nvim-neotest/neotest",
    optional = true,
    opts = {
      adapters = {
        ["rustaceanvim.neotest"] = {},
      },
    },
  },
}
```

</details>

**Comment by dpetka2001** [comment]
So, I started `bacon` in another terminal and started the `clippy` job by pressing `c` and it correctly shows diagnostics in `bacon`. However, in Neovim I don't see any diagnostics. Is there maybe some additional configuration needed on the user end?

**Comment by crisidev** [comment]
Bacon requires a specific configuration to work with the language server.

Have you followed the installation instructions here by any chance here https://github.com/crisidev/bacon-ls/blob/main/README.md?

**Comment by mirsella** [comment]
Personally no I didn't check further, sorry, as usually lsp are plug and play, especially in lazyvim.
i'm sure it's my who is having a problem, but even witht the readme i can't manage to make it work.
from the repro.lua i've given, i've just added the lazyvim config from you readme, and everything is still the same.
of course bacon and bacon-ls are both installed with mason from the start.

**Comment by crisidev** [comment]
I'll try to finish the PR early next week btw, the mason.nvim inclusion for bacon and bacon-ls was merged a couple of days ago.

**Comment by dpetka2001** [comment]
In the same directory where `Cargo.toml` resides, I run `bacon --init` and that created a default `bacon.toml` file. I then added at the end of that file 
```
[export]
enabled = true
path = ".bacon-locations"
line_format = "{kind}:{path}:{line}:{column}:{message}"
```
I left the `bacon-ls` configuration in `nvim-lspconfig` untouched with the default values. I then started bacon in another terminal and pressed `c` to start the clippy job. It shows correctly diagnostics. A file is created in the same directory as `Cargo.toml` and `bacon.toml` that is named `.bacon-locations` just like it's supposed to. However, when I open Neovim I still can't see any diagnostics. Are there any other steps that I'm maybe missing?

**Comment by mirsella** [comment]
Following @dpetka2001 instruction and my repro.lua i got it working

**Comment by dpetka2001** [comment]
@mirsella Seriously? How? Why is it not working for me?

**Comment by mirsella** [comment]
i was using mason's bacon and bacon-ls btw don't know if it can change anything  :shrug:

**Comment by crisidev** [comment]
Can you share the content of `.bacon-locations`  after you have run it in your repo?

I'll try to use the repro this weekend to check what's going on with your config, because "it works on my computer" 😞

**Comment by dpetka2001** [comment]
Funnily enough I just created a new Rust project with `cargo init` and there it works correctly as it should. So, there's no need for further help, really appreciate it. There was another change today for the Rust Extra, reflecting some changes in `rustaceanvim`, so maybe you would like to take that into account. Nothing big, one line really :stuck_out_tongue:

**Comment by crisidev** [comment]
This is ready for review now!

File: lua/lazyvim/plugins/extras/lang/rust.lua [hunk 1]
```diff
@@ -64,8 +80,14 @@ return {
                 enable = true,
               },
             },
-            -- Add clippy lints for Rust.
-            checkOnSave = true,
+            -- Add clippy lints for Rust if using rust-analyzer
+            checkOnSave = {
+              enable = diagnostics == "rust-analyzer",
+            },
```
[the remark this answers is not in the document]
Comment (crisidev) [reply lua/lazyvim/plugins/extras/lang/rust.lua:86]: Thanks, I though both worked. I'll fix it.

File: lua/lazyvim/plugins/extras/lang/rust.lua [hunk 2]
```diff
@@ -37,7 +47,13 @@ return {
   {
     "williamboman/mason.nvim",
     optional = true,
-    opts = { ensure_installed = { "codelldb" } },
+    opts = function(_, opts)
+      opts.ensure_installed = opts.ensure_installed or {}
+      vim.list_extend(opts.ensure_installed, { "codelldb" })
+      if diagnostics == "bacon-ls" then
+        vim.list_extend(opts.ensure_installed, { "bacon", "bacon-ls" })
```
Comment (dpetka2001) [inline lua/lazyvim/plugins/extras/lang/rust.lua:54]: You could also remove `bacon-ls` from `mason.ensure_installed`, since LSP servers that are defined in `nvim-lspconfig` spec get installed automatically if they are `enabled`. This is just a suggestion and doesn't mean you absolutely have to do it. It's just how the rest of LazyVim LSP servers are defined as well throughout the codebase. [Corresponding code snippet in LazyVim](https://github.com/LazyVim/LazyVim/blob/a33eabddd9f3786421652d07894516d946c88647/lua/lazyvim/plugins/lsp/init.lua#L218-L231).

Comment (crisidev) [reply lua/lazyvim/plugins/extras/lang/rust.lua:54]: `bacon` needs to be remain if I am not mistaken, since it's not an lsp server itself..

Comment (dpetka2001) [reply lua/lazyvim/plugins/extras/lang/rust.lua:54]: Yes correct.

**Comment by crisidev** [comment]
Is there anything else to address in this PR or can it be merged?

**Comment by dpetka2001** [comment]
Only the approval of the maintainer himself and maybe if he has some suggestions of his own. As I don't do Rust and am just a simple user like you, I have no say. It would also be nice if some people who actually do Rust share their experience with this PR, as they are more knowledgeable about possible pitfalls (if any) I'm not aware of.

**Comment by mirsella** [comment]
Already shared by opinions :smiley: https://github.com/LazyVim/LazyVim/pull/3212#issuecomment-2168771754

tldr: i wouldn't use bacon mainly because if I don't run the `bacon` program in another shell i won't have ANY diagnostics, because rust analyzer would be disabled. it means if i want to do a quick change/view anything in a rust project, instead of just opening my editor i also have to open a new tab and launch bacon, then launch my editor.
might not be a issue for other user tho, and using bacon-ls instead of RA is opt in anyway.

Good work OP on this PR and bacon in general !

**Comment by crisidev** [comment]
> Already shared by opinions 😃 [#3212 (comment)](https://github.com/LazyVim/LazyVim/pull/3212#issuecomment-2168771754)
> 
> tldr: i wouldn't use bacon mainly because if I don't run the `bacon` program in another shell i won't have ANY diagnostics, because rust analyzer would be disabled. it means if i want to do a quick change/view anything in a rust project, instead of just opening my editor i also have to open a new tab and launch bacon, then launch my editor. might not be a issue for other user tho, and using bacon-ls instead of RA is opt in anyway.
> 
> Good work OP on this PR and bacon in general !

Thanks for sharing this! I'll answer below.

**Comment by crisidev** [comment]
> Following @dpetka2001 instruction and my repro.lua i got it working
> 
> edit: it's my personal opinion, just throwing some first impression it is indeed faster, but rust_analyzer is still very good, and i'll probably stick with it for now:
> 
>     * need to have an extra file bacon.toml

This is something that needs to be done once, as `bacon.toml` can be shared between project if put in the right folder in your home.

> 
>     * need to put some extra config in it that i'll have to search on GitHub every time

Again, this needs to be done only once :)
> 
>     * need to launch bacon

I am happy to add launching `bacon` to `bacon-ls` to ease the onboarding pain.
What I am currently doing in my config is something along the lines of this:
```lua
            local function bacon_term()
                LazyVim.terminal.open(
                    { "bacon", "clippy", "--", "--all-features", "--target", vim.env.NEOVIM_RUST_TARGET },
                    {
                        ft = "bacon",
                        cwd = LazyVim.root.get(),
                        env = { LAZYTERM_TYPE = "bacon" },
                    }
                )
            end

            if vim.g.lazyvim_rust_diagnostics == "bacon-ls" then
                bacon_term()
                vim.defer_fn(function()
                    bacon_term()
                end, 2000)
                vim.keymap.set({ "n", "i", "t" }, "<c-y>", bacon_term, { desc = "Bacon" })
            end
```

> 
>     * if i'm on quick project/another person repo which most didn't setup theses things, i'll have 0 diagnostic because `vim.g.lazyvim_rust_diagnostics = "bacon-ls"` would still be set, so i'll need to setup everything for every project.
> 
> 
> It would need a few things for it to be usable for me: if the cargo project is set up for bacon + bacon-ls, then use it, and auto launch bacon if not running (the autostart can be a flag in bacon-ls) if the cargo project is not setup for bacon + bacon-ls, fallback to rust_analyzer

File: lua/lazyvim/plugins/extras/lang/rust.lua [hunk 3]
```diff
@@ -79,6 +99,14 @@ return {
       },
     },
     config = function(_, opts)
+      -- Configure dap with codelldb
+      local package_path = require("mason-registry").get_package("codelldb"):get_install_path()
+      opts.dap = {
+        adapter = require("rustaceanvim.config").get_codelldb_adapter(
+          package_path .. "/codelldb",
+          package_path .. "/extension/lldb/lib/liblldb.dylib"
```
Comment (crisidev) [inline lua/lazyvim/plugins/extras/lang/rust.lua:107]: Mmmm, now that I relook at this, it requires some changes to also work on Linux..

**Comment by mirsella** [comment]
ohhhh it can be configured globally, nice !

why not make it a extra btw ?

**Comment by mirsella** [comment]
ohhhh it can be configured globally, nice !

well i have nothing more to say. might try to use this and bacon when it's merged

**Comment by twilwa** [comment]
Wanted to check in here -- should I add bacon from rustup so I can run the bacon --init command in my home config? At the moment it's not in my PATH, although I've installed it in nvim, or are we waiting on this merge for the mason install to complete correctly (& handle the bacon initialization accordingly?)

**Comment by folke** [comment]
I'm willing to look at this, but there's currently merge conflicts

**Comment by crisidev** [comment]
I will solve the merge conflict next week. I am currently on holidays without a laptop 😄

**Comment by folke** [comment]
Enjoy!

**Comment by crisidev** [comment]
I fixed the conflicts and update the codelldb settings to support both .dylib and .so files. 

I'll do some testing to make sure it works properly.

**Comment by crisidev** [comment]
Testing done, it seems to be working properly.

**Comment by folke** [comment]
ty!
